flocker/volume/service.py

```diff
@@ -216,30 +183,76 @@ def wait_for_volume(self, name):
 
         Polls the storage pool for the specified volume to appear.
 
+        XXX: Poll Openstack  API until block is unattached.
+
         :param VolumeName name: The name of the volume.
 
         :return: A ``Deferred`` that fires with a :class:`Volume`.
         """
-        def check_for_volume(node_id, name):
-            d = self.enumerate()
-            d.addCallback(compare_volumes_by_name_node_id, node_id, name)
-            return d
-
-        def compare_volumes_by_name_node_id(volumes, node_id, name):
-            """
-            Iterate the volumes managed by this service and compare the
-            node ID and the ``VolumeName`` to determine if we have found the
-            ``Volume`` instance requested by name.
-            """
-            for volume in volumes:
-                if volume.node_id == node_id and volume.name == name:
-                    return volume
-            return deferLater(
-                self._reactor, WAIT_FOR_VOLUME_INTERVAL,
-                check_for_volume, node_id, name
-            )
-
-        return check_for_volume(self.node_id, name)
+        compute_driver, volume_driver = driver_from_environment()
+
+        openstack_volumes = volume_driver.list()
+        for openstack_volume in openstack_volumes:
+            if openstack_volume.name == name.to_bytes():
+                break
+
+        # Wait for volume to detach
+        # We need to know what the current node IP is here, or supply
+        # current node as an attribute of OpenstackStoragePool
+        public_ips = get_public_ips()
+        all_nodes = compute_driver.servers.list()
+        for node in all_nodes:
+            if ipaddr.IPv4Address(node.accessIPv4) in public_ips:
+                break
+
+        while True:
+            # Do we need this or is the availability check done on the server
+            # side?
+            openstack_volume = volume_driver.get(openstack_volume.id)
+            if openstack_volume.status == u'available':
+                break
+            time.sleep(0.5)
+
+        device_path = next_device()
+        # Sometimes this raises:
+        # Exception: 500 Server Error The server has either erred or is incapable of performing the requested operation.
+        openstack_volume.attach_to_instance(instance=node, mountpoint=device_path)
+
+        # Wait for device to appear
```

Comment:
Another significant delay while the device is recognised by the OS. Discuss and investigate what sort of device this is and why it's so slow to attach.